### ISSUE TITLE
visualize_attention.py

### ISSUE THREAD

**Issue by karthik78437** [karthik78437/karthik#8]
"""
visualize_attention.py

Load an attention model that returns attention weights (set attention_return_attention=True
when building model), run predict and visualize attention for a sample.
"""

import numpy as np
import matplotlib.pyplot as plt
import os
from models import build_attention_lstm
from data_generation import generate_complex_timeseries, create_sequences
from utils import train_test_split_sequences, scale_sequences

def visualize_sample_attention(model, X_sample, y_sample, scaler=None, seq_index=0, savepath=None):
    """
    Given a model that outputs [prediction, attention_weights], plot the input series and barplot of attention weights.
    """
    # model.predict returns [preds, att_weights] shape (n_samples, 1) and (n_samples, time_steps)
    preds, att = model.predict(X_sample)
    pred = preds[seq_index, 0]
    att_seq = att[seq_index]  # shape (time_steps,)

    # retrieve target series from X_sample
    series = X_sample[seq_index, :, 0]  # assuming target is feature 0 used as past
    time = np.arange(len(series))

    plt.figure(figsize=(12, 4))
    plt.plot(time, series, label='input target (past)')
    plt.axhline(y=y_sample[seq_index], color='r', linestyle='--', label='true next')
    plt.scatter([len(series)], [pred], color='k', label='predicted next')
    plt.legend()
    plt.title('Input series and prediction')

    plt.figure(figsize=(12, 3))
    plt.bar(np.arange(len(att_seq)), att_seq)
    plt.xlabel('Time step (older -> newer)')
    plt.ylabel('Attention weight')
    plt.title('Attention weights across time steps')
    if savepath:
        plt.savefig(savepath)
    plt.show()

if __name__ == '__main__':
    # Quick demo (requires you to have trained a model that returns attention)
    df = generate_complex_timeseries(n_steps=1200, n_exog=3)
    X, y = create_sequences(df, seq_len=48)
    X_tr, X_te, y_tr, y_te = train_test_split_sequences(X, y, test_size=0.2)
    X_tr_s, X_te_s, scaler = scale_sequences(X_tr, X_te)

    input_shape = (X_tr_s.shape[1], X_tr_s.shape[2])
    model = build_attention_lstm(input_shape, lstm_units=32, attention_return_attention=True)
    # load weights if available: model.load_weights('path_to_saved_weights.h5')
    print("Model built; to visualize, load weights from trained model and call visualize_sample_attention(...)")
